Get the current Agent container build working again

portante opened the pull request:
Replaces PR #3013.

Apparently, the Agent container builds aren't working any more.  This PR restores them to working order.

The principal changes:

- The container build depends on the `pbench.repo.j2` template from the Ansible collection...which, apparently, moved recently (#2991)
- Added a second repository to the `agent/containers/images/repo.yml.j2` template for the "auxillary" RPMs

We also made some smaller changes:
- Added Fedora-36 to the list of default targets
- The `-TEST` suffix will now _really_ disappear when it's not needed
- Removed some redundant dependencies on the `pkgmgr-clean` target
- Tweaked a code comment
- Removed an unnecessary PCP repo enablement
- Enabled a method of quickly executing all `make` targets for container build

dbutenhof commented on agent/containers/images/apply-tags:
```diff
@@ -13,7 +13,7 @@ fi
 
 function apply_tag {
-    buildah tag ${1}:${githash} ${1}:${2}
+    ${BUILDAH_ECHO} buildah tag ${1}:${githash} ${1}:${2}
```
Where does `BUILDAH_ECHO` come from (I don't see it in `main`)? I assume the intent is something along the lines of `BUILDAH_ECHO=echo make` to disable the container commands for testing, but a descriptive comment somewhere would seem appropriate.

webbnh replied:
As you surmise, it is intended to allow for fast-turnaround, low-impact testing.

By default, `BUILDAH_ECHO` is undefined, and so it expands vacuously with no effect, leaving the `buildah` commands to execute normally; however, if it is defined to `echo`, then, instead of executing the commands, they are simply echoed.

I think there is a better way to accomplish this, but it would be more effective if the `buildah` commands were placed in the `make` recipes instead of wrapped inside external scripts; so, for the moment, this approach is worth adopting.

dbutenhof replied:
I'd prefer to have this at least recorded in a comment somewhere, because it looks rather weird. But "whatever" ... I'll resolve this.

webbnh reviewed (approved):
👍 

(@ndokos, does this work in your environment?)

ndokos reviewed (dismissed):
LGTM - I don't get the Makefile errors I was getting and I was able to build a set of images correctly.

I may have been misreading the README file but it seems  that e.g. `make fedora-36` works but `make fedora-36-all` does not do the major/minor tagging. @webbnh pointed out that the block comment at the beginning of the Makefile mentions the former. The README file says:

```
There are make targets for each of the four supported distributions,
CentOS 8 (`centos-8`), CentOS 7 (`centos-7`), Fedora 33 (`fedora-33`),
and Fedora 32 (`fedora-32`).  There are also make targets for each
subset of the four images (all, base, tools, workloads) built for
each distribution, e.g. `centos-8-tools`, `fedora-32-base`, etc.
```

Perhaps the `all` in the last list should be deleted?

webbnh commented:
> Perhaps the `all` in the last list should be deleted?

The problem is that there is an asymmetry in the target-sets:
- building `tds` or `centos-8-tds` depends on `centos-8-tool-data-sink-tagged`
  - `centos-8-tool-data-sink-tagged` depends on `centos-8-tool-data-sink`
- building `all` or `centos-8` depends on `centos-8-all-tagged`
  - `centos-8-all-tagged` depends on `centos-8-all`

So, if one requests the `centos-8-all` target (thinking that it's an analog of the `centos-8-tds` target), then it skips the other dependences of the `centos-8` and `centos-8-all-tagged` targets.

That is, I think the problem is that the `%-all` target is misnamed:  it should be `%-all-things` or the equivalent, and we should have a different `%-all` target which is a proper peer of the `%-tds` and `%-tm` targets.

I.e., line 130 should be (so that it matches either `centos-8` or `centos-8-all`:
```
${_DISTROS} $(_DISTROS:%=%-all): %: all-tags %-all-tagged
```
and lines 184-187 should be (so that we rename the `%-all` target to `%-all-things`:
```
%-all-tagged: %-all-things %-tags.lis
	...
%-all-things: %-workloads-tagged %-tool-data-sink-tagged %-tool-meister-tagged %-all.Dockerfile %-tags.lis
```
(although, we might want a word other than "things"...like "components" or something 😉).

webbnh commented:
> it should be `%-all-things` or the equivalent, and we should have a different `%-all` target which is a proper peer of the `%-tds` and `%-tm` targets.

However, perhaps this would be a good change for a separate PR.

webbnh commented on agent/containers/images/Makefile:
```diff
@@ -127,11 +127,11 @@ tm: pkgmgr-clean all-tags $(_DEFAULT_DISTROS:%=%-tool-meister-tagged)
-${_DISTROS}: %: pkgmgr-clean %-all-tagged
+${_DISTROS}: %: all-tags %-all-tagged
 
-$(_DISTROS:%=%-tds): %-tds: pkgmgr-clean %-tool-data-sink-tagged
+$(_DISTROS:%=%-tds): %-tds: all-tags %-tool-data-sink-tagged
 
-$(_DISTROS:%=%-tm): %-tm: pkgmgr-clean %-tool-meister-tagged
+$(_DISTROS:%=%-tm): %-tm: all-tags %-tool-meister-tagged
```
This change is not a good idea:  it will prevent us from building a distro which is not in the default list.

These targets are intended to build for a single distro.  As such, they don't need (or want!) to depend on `all-tags`, because the single distro has no requirement to be consistent with any other distro.

This change should be reverted.

webbnh replied:
> it will prevent us from building a distro which is not in the default list.

OK, it won't actually _prevent_ it -- what will happen is that `make` will go and generate the `*-tag.lis` files for each of the default distros, make sure that they all match each other, and then proceed to ignore them and generate the `*-tag.lis` file for the distro that it actually intends to build (and it won't compare it to the others).  Unfortunately, generating the `*-tag.lis` files is slow, so it's painful to be doing it for no purpose.

Moreover, if for some reason one of the default `*-tag.lis` files cannot be built, then the target won't be built either, even though it doesn't depend on the defaults' files.  (Obviously, this can be worked around by overriding the default distro list, but that shouldn't be have to be necessary.)

webbnh reviewed (changes requested):
On subsequent review....

portante commented:
> LGTM - I don't get the Makefile errors I was getting and I was able to build a set of images correctly.

Great to hear!

> I may have been misreading the README file but it seems that e.g. `make fedora-36` works but `make fedora-36-all` does not do the major/minor tagging.

Right.  Because we have the `all` Makefile target and the `all` container image kind or type.  I have replaced the `all` target with `everything` to avoid this confusion.

> @webbnh pointed out that the block comment at the beginning of the Makefile mentions the former. The README file says:
> 
> ```
> There are make targets for each of the four supported distributions,
> CentOS 8 (`centos-8`), CentOS 7 (`centos-7`), Fedora 33 (`fedora-33`),
> and Fedora 32 (`fedora-32`).  There are also make targets for each
> subset of the four images (all, base, tools, workloads) built for
> each distribution, e.g. `centos-8-tools`, `fedora-32-base`, etc.
> ```
> 
> Perhaps the `all` in the last list should be deleted?

See above and I have updated the list of image kinds.

dbutenhof reviewed (approved):
OK, we've gone back and forth on the "approve with comments" thing, but while I like the technical changes and I'm willing to go with this I also have some minor formatting and wording issues. If you don't want to change things, I'll resolve the comments.

dbutenhof reviewed (dismissed):
Yeah, I like exposing the Makefile documentation in the README. Although I kinda miss having it inline, it's a big and unwieldy comment block that's arguably more appropriate and digestible in the README format.

webbnh commented on agent/containers/images/README.md:
```diff
@@ -52,57 +51,107 @@ localhost/pbench-agent-all-centos-8   v0.69.3-1  9396f0337681 ...
-for each distribution, e.g. `centos-8-tools-tagged`,
-`fedora-35-base-tagged`, etc.
+There are make targets for each of the five supported distributions, CentOS 9
+(`centos-9`), CentOS 8 (`centos-8`), CentOS 7 (`centos-7`), Fedora 33
+(`fedora-33`), and Fedora 35 (`fedora-35`).  There are also make targets for
```
That should be "Fedora _36_", not 33, right?

webbnh commented on agent/containers/images/Makefile:
```diff
@@ -108,7 +83,7 @@ _ALL_RPMS = ${_TOOL_RPMS} ${_WORKLOAD_RPMS}
 
 # By default we build the Tool Data Sink and Tool Meister images as well.
-all: all-tags $(_DEFAULT_DISTROS:%=%-all-tagged)
+everything: all-tags $(_DEFAULT_DISTROS:%=%-all-tagged)
```
How do this change improve anything?  (AFAIK, `make all` is a more common idiom than `make everything`...here's [one set of conventions](https://www.gnu.org/software/make/manual/html_node/Standard-Targets.html#Standard-Targets)....)

This change doesn't address the problem which [I outlined above](https://github.com/distributed-system-analysis/pbench/pull/3018#issuecomment-1258434255), which is with the `%-all` target, not the default, `all` target.

portante replied:
As I wrote in the commit message, we have a container kind or type named `all`.  It seemed to me that that caused all this confusion.  So by renaming the `all` target to `everything` we remove that confusion.

webbnh replied:
The problem seems to stem from the conflict between the `make` conventions regarding the token `all` and the use of that token to name one of the containers in the distro.  Since it will be disruptive to rename that container, we're left either with bucking the `make` conventions or having an inconsistency between the mid-level targets and the artifacts that they build.

Changing the top-level target is undesirable, but it seems to be the least of evils.  The default target doesn't need to be explicitly specified, so having it named `everything` instead of `all` is not a huge problem (since people can just omit it), particularly now that the Makefile will object to the use of the `all` target.

webbnh commented on agent/containers/images/Makefile:
```diff
@@ -127,11 +102,11 @@ tm: pkgmgr-clean all-tags $(_DEFAULT_DISTROS:%=%-tool-meister-tagged)
-${_DISTROS}: %: pkgmgr-clean %-all-tagged
+${_DISTROS}: %: %-all-tagged
 
-$(_DISTROS:%=%-tds): %-tds: pkgmgr-clean %-tool-data-sink-tagged
+$(_DISTROS:%=%-tds): %-tds: %-tool-data-sink-tagged
 
-$(_DISTROS:%=%-tm): %-tm: pkgmgr-clean %-tool-meister-tagged
+$(_DISTROS:%=%-tm): %-tm: %-tool-meister-tagged
```
Is there some reason why these targets should _not_ depend on the `pkgmgr-clean` target?

(That dependency was replaced with `all-tags` which in turn depends on `pkgmgr-clean`, but now that we're reverting that change, shouldn't we be restoring the original dependency?)

portante replied:
This was changed from `all-tags` to `pkgmgr-clean` in PR #2607 with no particular comment that I could find describing the change.  We have `pkgmgr-clean` as a dependency of `all-tags` to help ensure all the distributions are using the same pbench-agent version.  Why we need this per distribution, where the version for that distribution is self-consistent, and not each time we generate a `*-tags.lis` file is still unclear.  I had removed it because of that inconsistency.

I'll replace it in order for this PR to move forward and somebody else can solve that problem in a future PR.

webbnh replied:
The `all-tags` dependency was removed because it prevented these targets from building when there was a problem with any of the default distros on COPR.  (E.g., we couldn't build Fedora-36 unless we had a RHEL-7 RPM.)

Previously, these targets depended on `pkgmgr-clean` indirectly via their dependency on `all-tags`; so, when I removed the `all-tags` dependency, I replaced it with `pkgmgr-clean` to maintain that dependency.

I still don't know why we need to clean the dnf cache before building the `*-tags.lis` files, but it seems like, if we need to do it before doing a multiple-distro build, then we should need to do it before doing a single-distro build...but maybe I'm just confused.

webbnh commented on agent/containers/images/README.md:
```diff
@@ -52,57 +51,107 @@ localhost/pbench-agent-all-centos-8   v0.69.3-1  9396f0337681 ...
+ * `<DISTRO>-push-<TYPE>` (e.g., "fedora-34-push-alpha"):  push the specified
+   containers
+
+_NOTE_: the supported distributions are centos-# and fedora-#, where the
+minimum CentOS version is 7, and the minimum Fedora version is the latest
+non-end-of-life releases.
```
This isn't completely true (depending on the details of the meaning of "supported", of course).  The user can conceivably build for any distribution for which they have access to the appropriate RPMs, although there is some "special handling" for particular versions which may or may not be adequate for "unsupported" versions.

I don't think it makes good sense to mix general and specific, here.  I think we should either list exactly which distro/version pairs we support, or we should make a more general statement of our intent (and let the user discover for themself whether we've keeping up).

I might be inclined to point the reader at the definition of `_DEFAULT_DISTROS` in the makefile, as that's possibly the most concrete expression of support that we've got.

webbnh reviewed (changes requested):
I think it's a bad idea to banish the interface documentation from the `Makefile`.  Yeah, it's big, but that's because the interface is big; I don't think it actually causes a problem to have it where it was, and I think it is beneficial.  But, if you both think it's better, I'm willing to bow to your wisdom.

However, I think changing the default target from `all` to `everything` is counterproductive, and it doesn't address the problem which I pointed out above.

Also, the change to the top-level distro-specific targets was not completely reverted...was that intentional?

Finally, there's a key typo in the docs and a note which I think should be improved.

portante commented on agent/containers/images/Makefile:
```diff
@@ -1,31 +1,6 @@
-#     pkgmgr-clean:  clear the local package manager cache
-#     all-tags:  build all default distro "-tags.lis" files and verify that they are consistent.
-#     all-dockerfiles:  build all default distro ".repo" and ".Dockerfile" files
+# See the README.md file for a description of the interface.
```
It doesn't matter to me.  If @ndokos likes it the other way, I'll drop the patch sets from this PR.

ndokos replied:
Ideally, what I would like to have is `make help` which prints out the most common set of targets  (or examples of such)  in less than a "page" and refers to wherever the complete documentation can be found (whether as a block comment in the Makefile or in the README does not make much difference). SInce I knew that the README existed, that's where I went to find what I needed to find - I did not think to look at the block comment until @webbnh pointed it out.

dbutenhof replied:
Well, since the default `all` target just prints that you shouldn't do that, maybe it should depend on a `help` target instead which prints an extended (but suitably brief) explanation. In fact, ideally this might be a `MAKE.md` that could stand along from the overall `README.md` (which could link to `MAKE.md`). How easy is it to format markdown on the command line? 🤔

webbnh replied:
I think having a `help` target would be lovely, particularly if it were "automagic" (e.g., generated by introspection, or something).  However, the notion of having to keep _three_ bits of text up to date (the help text, the "module" header, and the external READ-ME file), is so unsatisfying.  😝  That means, at best, the `help` target would probably just refer the user to the `README.md` file, but that's problematic -- how is the user supposed to format it for human consumption, and what happens if the file is missing?  😞

webbnh commented on agent/containers/images/Makefile:
```diff
@@ -102,11 +106,11 @@ tm: all-tags $(_DEFAULT_DISTROS:%=%-tool-meister-tagged)
+${_DISTROS}: %: pkgmgr-clean %-all-tagged
 
-$(_DISTROS:%=%-tds): %-tds: %-tool-data-sink-tagged
+$(_DISTROS:%=%-tds): %-tds: pkgmgr-clean %-tool-data-sink-tagged
```
Shall we add
```
${_DISTROS:%=%-everything}: %-everything: pkgmgr-clean %-all-tagged
```
to enable (e.g.) `centos-8-everything` as a target (analogous to the `centos-8-tds` target, as the `everything` target is to the `tds` target)?

And, if you want to be DRYer after adding that line, change line 109 to:
```
${_DISTROS}: %: %-everything
```

webbnh reviewed (approved):
One @ndokos settles the documentation argument (🤣), we're good by me, unless you want to accept my suggestion below.

ndokos reviewed (approved):
LGTM - I have a slight preference for the docs to be complete in the README. As I mentioned, I would like a `make help` with the most commonly used targets, but that's for a later PR.

webbnh reviewed (approved):
👍
